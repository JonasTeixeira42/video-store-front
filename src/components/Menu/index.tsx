import React, { useState, useCallback } from 'react';
import { Menu3 as MenuIcon } from '@styled-icons/remix-line/Menu3';
import { Close as CloseIcon } from '@styled-icons/material-outlined/Close';

import Logo from 'components/Logo';
import MediaMatch from 'components/MediaMatch';
import ModalAddMovies from 'components/ModalAddMovies';
import ModalAddDirectors from 'components/ModalAddDirectors';

import * as S from './styles';

type MenuProps = {
  onCreateMovie?: () => void;
};

const Menu = ({ onCreateMovie }: MenuProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const [openAddMovie, setOpenAddMovie] = useState(false);
  const [openAddDirector, setOpenAddDirector] = useState(false);

  const toggleAddModal = useCallback(() => {
    setOpenAddMovie(!openAddMovie);

    !!onCreateMovie && onCreateMovie();
  }, [openAddMovie, onCreateMovie]);

  const toggleDirectorModal = useCallback(() => {
    setOpenAddDirector(!openAddDirector);
  }, [openAddDirector]);

  return (
    <>
      {openAddMovie && (
        <ModalAddMovies isOpen={openAddMovie} setIsOpen={toggleAddModal} />
      )}
      {openAddDirector && (
        <ModalAddDirectors
          isOpen={openAddDirector}
          setIsOpen={toggleDirectorModal}
        />
      )}
      <S.Wrapper>
        <Logo size="normal" hiddenOnMobile />

        <MediaMatch greaterThan="medium">
          <S.MenuNav>
            <S.MenuOption onClick={toggleAddModal}>Add Movie</S.MenuOption>
            <S.MenuOption onClick={toggleDirectorModal}>
              Add Director
            </S.MenuOption>
          </S.MenuNav>
        </MediaMatch>

        <MediaMatch lessThan="medium">
          <S.IconWrapper onClick={() => setIsOpen(true)}>
            <MenuIcon aria-label="Open Menu" />
          </S.IconWrapper>
        </MediaMatch>

        <S.FullMenu aria-hidden={!isOpen} isOpen={isOpen}>
          <CloseIcon aria-label="Close Menu" onClick={() => setIsOpen(false)} />

          <S.MenuNav>
            <S.MenuOption onClick={toggleAddModal}>Add Movie</S.MenuOption>
            <S.MenuOption onClick={toggleDirectorModal}>
              Add Director
            </S.MenuOption>
          </S.MenuNav>
        </S.FullMenu>
      </S.Wrapper>
    </>
  );
};

export default Menu;
