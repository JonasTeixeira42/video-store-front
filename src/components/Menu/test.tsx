import { fireEvent, screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/tests/helpers';

import Menu from '.';

describe('<Menu />', () => {
  it('should render the menu', () => {
    renderWithTheme(<Menu />);

    expect(screen.getByLabelText(/open menu/i)).toBeInTheDocument();
    expect(screen.getAllByText('Add Movie')[0]).toBeInTheDocument();
    expect(screen.getAllByText('Add Director')[0]).toBeInTheDocument();
    expect(
      screen.getByRole('img', { name: /Clapperboard/i }),
    ).toBeInTheDocument();
  });

  it('should handle the open/close menu', () => {
    renderWithTheme(<Menu />);

    const fullMenu = screen.getByRole('navigation', { hidden: true });

    expect(fullMenu.getAttribute('aria-hidden')).toBe('true');
    expect(fullMenu).toHaveStyle({ opacity: 0 });

    fireEvent.click(screen.getByLabelText(/open menu/i));
    expect(fullMenu.getAttribute('aria-hidden')).toBe('false');
    expect(fullMenu).toHaveStyle({ opacity: 1 });

    fireEvent.click(screen.getByLabelText(/close menu/i));
    expect(fullMenu.getAttribute('aria-hidden')).toBe('true');
    expect(fullMenu).toHaveStyle({ opacity: 0 });
  });

  it('should open Add Movie Modal', () => {
    renderWithTheme(<Menu />);

    fireEvent.click(screen.getAllByText('Add Movie')[0]);

    expect(
      screen.getByRole('heading', { name: /Add a Movie/i }),
    ).toBeInTheDocument();
  });

  it('should open Add Director Modal', () => {
    renderWithTheme(<Menu />);

    fireEvent.click(screen.getAllByText('Add Director')[0]);

    expect(
      screen.getByRole('heading', { name: /Add a Director/i }),
    ).toBeInTheDocument();
  });
});
