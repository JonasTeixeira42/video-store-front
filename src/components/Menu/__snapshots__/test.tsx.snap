// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should render the menu 1`] = `
.c12 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  color: #FAFAFA;
}

.c2 .{
  width: 4.5rem;
  height: 4.5rem;
}

.c2 .c3 {
  font-size: 3.2rem;
  left: calc(4.5rem + 0.5rem);
  bottom: -1.2rem;
}

.c4 {
  position: absolute;
}

.c5 {
  display: none;
}

.c10 {
  display: none;
}

.c0 .sc-jrsJCI {
  margin-bottom: 3.2rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 2.4rem 0;
  position: relative;
}

.c11 {
  color: #FAFAFA;
  width: 2.4rem;
  height: 2.4rem;
  cursor: pointer;
}

.c7 {
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
}

.c9 {
  position: relative;
  color: #FAFAFA;
  font-size: 1.8rem;
  margin: 0.3rem 2.4rem 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  display: inline-block;
  cursor: pointer;
}

.c9:hover::after {
  content: '';
  position: absolute;
  display: block;
  height: 0.3rem;
  background: #F231A5;
  -webkit-animation: hoverAnimation 0.2s forwards;
  animation: hoverAnimation 0.2s forwards;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background: #FAFAFA;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  height: 100vh;
  overflow: hidden;
  -webkit-transition: opacity 0.3s ease-in-out;
  transition: opacity 0.3s ease-in-out;
  opacity: 0;
  pointer-events: none;
}

.c13 > svg {
  position: absolute;
  top: 0;
  right: 0;
  margin: 1.6rem;
  cursor: pointer;
  width: 2.4rem;
  height: 2.4rem;
}

.c13 .c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c13 .c8 {
  color: #030517;
  font-weight: 600;
  font-size: 2.0rem;
  margin-bottom: 2.4rem;
  -webkit-transform: translateY(3rem);
  -ms-transform: translateY(3rem);
  transform: translateY(3rem);
  -webkit-transition: -webkit-transform 0.3s ease-in-out;
  -webkit-transition: transform 0.3s ease-in-out;
  transition: transform 0.3s ease-in-out;
}

@media (max-width:768px) {
  .c2 {
    height: 4.5rem;
  }

  .c2 .{
    width: 4.5rem;
    height: 4.5rem;
  }

  .c2 .c3 {
    display: none;
    pointer-events: none;
  }
}

@media (min-width:768px) {
  .c5 {
    display: block;
  }
}

@media (max-width:768px) {
  .c10 {
    display: block;
  }
}

@media (min-width:768px) {
  .c7 {
    margin-left: 2.4rem;
  }
}

@media (max-width:768px) {
  .c7 {
    -webkit-order: 1;
    -ms-flex-order: 1;
    order: 1;
  }
}

<body>
  <div>
    <div
      aria-hidden="true"
      class="c0"
    />
    <div
      aria-hidden="true"
      class="c0"
    />
    <menu
      class="c1"
    >
      <div
        class="c2"
        color="white"
      >
        <div
          class=""
        >
          <svg
            aria-label="Clapperboard"
            fill="none"
            role="img"
            viewBox="-31 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="currentColor"
              fill-rule="evenodd"
            >
              <path
                d="M440.59 206.676H99.418l327.7-94.93a10.018 10.018 0 005.976-4.781 9.989 9.989 0 00.847-7.606L416.793 40.16C409.941 16.516 387.926 0 363.253 0c-5.198 0-10.378.738-15.401 2.191L40.176 91.321c-14.23 4.12-26.024 13.581-33.215 26.632-7.188 13.05-8.875 28.078-4.754 42.305l16.754 57.836v238.254C18.96 487.035 43.926 512 74.609 512h120.164c5.524 0 10-4.477 10-10s-4.476-10-10-10H74.613c-19.66 0-35.652-15.992-35.652-35.652V320.262H430.589v136.086c0 19.66-15.991 35.652-35.651 35.652H274.773c-5.52 0-10 4.477-10 10s4.48 10 10 10h120.165c30.687 0 55.652-24.965 55.652-55.652V216.676c0-5.524-4.477-10-10-10zm-176.332 93.586l42.488-73.586h55.262l-42.485 73.586zm-78.36 0l42.489-73.586h55.261l-42.484 73.586zm-78.355 0l42.484-73.586h55.266l-42.488 73.586zm37.18-129.457l-71.149-68.336 53.309-15.442a9.92 9.92 0 001.312 1.543l71.149 68.336-53.309 15.442a9.951 9.951 0 00-1.312-1.543zm134-125.84L349.87 113.3l-53.308 15.441a9.742 9.742 0 00-1.313-1.543l-71.148-68.336 53.308-15.441a9.92 9.92 0 001.313 1.543zM203.457 66.77l71.148 68.332-53.308 15.445a9.742 9.742 0 00-1.313-1.543l-71.148-68.336 53.309-15.441c.378.543.816 1.062 1.312 1.543zm149.961-45.368c3.21-.93 6.52-1.402 9.836-1.402 15.824 0 29.937 10.578 34.328 25.727l14.367 49.59-40.12 11.62a10.165 10.165 0 00-1.317-1.542l-71.145-68.333zM24.48 127.602c4.61-8.372 12.16-14.434 21.262-17.07l5.875-1.704a10.17 10.17 0 001.313 1.543l71.148 68.336-88.293 25.578-14.367-49.59c-2.637-9.097-1.547-18.718 3.062-27.093zm14.48 99.074h87.974l-42.485 73.586H38.961zm303.657 73.586l42.485-73.586h45.488v73.586zm0 0"
              />
              <path
                d="M303.922 405.113a9.997 9.997 0 00-5-8.66l-87.856-50.723a10.006 10.006 0 00-10 0 9.997 9.997 0 00-5 8.66v101.446a9.995 9.995 0 005 8.656 9.983 9.983 0 0010 0l87.856-50.719a9.997 9.997 0 005-8.66zm-87.856 33.403V371.71l57.856 33.402zm0 0M234.773 492c-5.507 0-10 4.492-10 10s4.493 10 10 10c5.512 0 10-4.492 10-10s-4.488-10-10-10zm0 0"
              />
            </g>
          </svg>
        </div>
        <p
          class="c3 c4"
        >
          Locadora
        </p>
      </div>
      <div
        class="c5"
      >
        <div
          class="c6 c7"
        >
          <div
            class="c8 c9"
          >
            Add Movie
          </div>
          <div
            class="c8 c9"
          >
            Add Director
          </div>
        </div>
      </div>
      <div
        class="c10"
      >
        <div
          class="c11"
        >
          <svg
            aria-hidden="true"
            aria-label="Open Menu"
            class="c12"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M3 4h18v2H3V4zm6 7h12v2H9v-2zm-6 7h18v2H3v-2z"
            />
          </svg>
        </div>
      </div>
      <nav
        aria-hidden="true"
        class="c13"
      >
        <svg
          aria-hidden="true"
          aria-label="Close Menu"
          class="c12"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
          />
        </svg>
        <div
          class="c6 c7"
        >
          <div
            class="c8 c9"
          >
            Add Movie
          </div>
          <div
            class="c8 c9"
          >
            Add Director
          </div>
        </div>
      </nav>
    </menu>
  </div>
  <div
    class="ReactModalPortal"
  />
  <div
    class="ReactModalPortal"
  />
</body>
`;
